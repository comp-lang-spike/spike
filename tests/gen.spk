#!/usr/bin/env spike

extern cdecl int printf() {}
extern cdecl int puts() {}
repr(o) { return o.printString; }

generator(gn) {
    obj i;
    i = 2;
    if (gn == 0) {
        return [yield 1; yield 2; yield 3; /*return 42;*/ while(true) yield i+=2;];
    } else if (gn == 1) {
        return [yield i+= 2; i];
    }
    return [i+= 2];
}

main(argv) {
    obj gn, g, i, z;
    
    z = Zoo.new();
    z = z yield: "airplane";
    puts(z);
    
    for (gn = 0; gn < 3; ++gn) {
        puts("generate!");
        g = generator(gn);
        for (i = 0; i < 10; ++i) {
            printf("%s\n", repr(g()));
        }
        puts("");
    }
    
    obj b;
    b = down();
    printf("back from hell: %s\n", repr(b()));
    return 0;
}

down() {
    obj b;
    b = [while (true) { yield 666; yield 42; } /*return b; cannotReenterBlock */ ];
    downDown(b);
    return b;
}

downDown(b) {
    obj i;
    for (i = 0; i < 10; ++i) {
        printf("pit of hell: %s\n", repr(b()));
    }
}


class Zoo {
    self yield: tomato {
        return tomato;
    }
}
