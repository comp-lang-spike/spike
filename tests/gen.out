airplane
generate!
1
2
3
4
6
8
10
12
14
16

generate!
4
4
6
6
8
8
10
10
12
12

generate!
4
6
8
10
12
14
16
18
20
22

pit of hell: 666
pit of hell: 42
pit of hell: 666
pit of hell: 42
pit of hell: 666
pit of hell: 42
pit of hell: 666
pit of hell: 42
pit of hell: 666
pit of hell: 42
back from hell: 666
